### PR TITLE
Add support for newer Django versions, up to 5.0

### PR DESCRIPTION
**Pull request by ZbigniewRA**
This PR builds on top of PR #800, which adds support for Django 4.1, 4.2, and Python 3.11.

My additions:
- Added support for newer Django versions, up to 5.0.
- Added support for Python 3.12.
- SHA1PasswordHasher is removed since Django 5.1, so was removed from the test application.

### CODE REVIEW

**Comment by johnthagen** [comment]
@ad-m would you mind reviewing this PR?

**Comment by GabrielCastelo-31** [comment]
Will the Django-Guardian project still be maintained? I think that this PR is really important to keep the project alive.

**Comment by johnthagen** [comment]
@GabrielCastelo-31 See

- #603

There has been some recent movement to give new maintainers permissions to move the project forward.

**Review by GabrielCastelo-31 — APPROVED** [review]
LGTM :+1:

File: .github/workflows/tests.yml
Comment (thclark) [inline .github/workflows/tests.yml:21]: Surely the `4.2` tag installs django `^4.2`, which would be `4.2.12` as of the time of writing. So I think we don't need this exclusion?

**Review by thclark — APPROVED** [review]
This is a really great PR, loads of attention to detail has been paid. Thanks so much for your work, folks!

**Comment by johnthagen** [comment]
@cmaggiulli This would be a great PR to review.

**Comment by cmaggiulli** [comment]
> a
Excellent will do

**Comment by cmaggiulli** [comment]
@johnthagen excellent will do.  I will add my review by earlier next week. At a high level - are we all OK with dropping support for the versions of Django and python outlined in the MR?

**Comment by johnthagen** [comment]
> Is the thought that this will be coupled with other MRs/features and eventually deployed as a new major release?

@cmaggiulli My vote is, yes. It's important to `django-guardian`'s long term health to add support for new Python and Django versions and drop support for end of life versions.

The fact there has not been a release in 3 years is another threat to its long term health because it will dissuade any new users from taking up the project.

In my opinion, the next short term goal should be releasing a new version with this MR and maybe other small MRs just to signal that it is alive again.

**Review by thclark — APPROVED** [review]

**Comment by cmaggiulli** [comment]
This hasnt fallen off my radar - Im just running a day or two behind

File: example_project/settings.py
Comment (cmaggiulli) [inline example_project/settings.py:78]: I realize this is mostly irrelevant since this is an example app, and this MR does not add this hasher, but I would suggest we change this to use PBKDF2PasswordHasher which has been available from 3.2 onward.  Django lists this as not recommended as far back as 3.2 [not recommended](https://github.com/django/django/blob/c3d3af8ea369e16263c8330b1745ba101d1c8d62/django/contrib/auth/hashers.py#L703)

File: guardian/testapp/migrations/0006_auto_20230727_0658.py
Comment (cmaggiulli) [inline guardian/testapp/migrations/0006_auto_20230727_0658.py:6]: At some point we should consider using a mix of both UUID primary keys and BigInt ( any non-char pks ) as we have experienced issues both in django proper and guardian upgrades with UUID fields.

File: example_project/requirements.txt
Comment (cmaggiulli) [inline example_project/requirements.txt:1]: should we =3.2.20,<6

File: setup.py
Comment (cmaggiulli) [inline setup.py:31]: may want to consider `install_requires=["Django>=3.2,<=5.0"]`

File: setup.py
Comment (cmaggiulli) [inline setup.py:15]: may want to consider `python_requires='>=3.8,<=3.12'`

File: setup.py
Comment (cmaggiulli) [inline setup.py:49]: 3.9 listed twice

File: tox.ini
Comment (cmaggiulli) [inline tox.ini:2]: downloadcache is entirely deprecated and has no effect.  I propose we remove

File: tox.ini
Comment (cmaggiulli) [inline tox.ini:8]: My 0.02 but unless we intent on regularly testing against django's devel Im not sure if I would keep `{core,example,docs}-py{310,311,312}-djangomain`.  Not now - just a thought going forward.  I realize this is out of scope for this MR

**Review by cmaggiulli — APPROVED** [review]
Looks good to me.  I would like to add some test cases for uuid fields and a mix of direct and generic foreign key tables.  That was my biggest concern here.  I played around with the code a bit in our environment and it seemed to work.  Building those tests were more time consuming that I originally anticipated so Ill add afterwards.  All my comments are suggestions - nothing impactful

File: example_project/requirements.txt
Comment (johnthagen) [reply example_project/requirements.txt:1]: @cmaggiulli In general capping versions is often not recommended as it removes the end consumer from being able to, for example, use an older django gaurdian that works fine with new Django but has not yet been tested at the time.

I think the general consensus is that it's better to leave these open and let the end consumer pin upper bounds in applications if needed but leave libraries more permissive. Especially for a library like Django that has stable backwards compatibility story.

File: setup.py
Comment (johnthagen) [reply setup.py:31]: Same thought in https://github.com/django-guardian/django-guardian/pull/816#discussion_r1682714364

I would avoid upper bounds.

File: setup.py
Comment (johnthagen) [reply setup.py:15]: For similar reasons to https://github.com/django-guardian/django-guardian/pull/816#discussion_r1682714364, I would recommend that libraries like this not cap and upper bound on Python versions or dependencies (unless absolutely necessary).

`django-guardian` is a pure Python package with no binary wheels, so I think it's even less useful to pin upper Python bounds.

Comment (cmaggiulli) [reply setup.py:15]: Good to know - I actually didn't know that guardian was not packaged and distributed as a whl.  Thanks for the info!

Comment (johnthagen) [reply setup.py:15]: It's packaged as a wheel, it's just a pure Python wheel (rather than any native complied code), thus should work fine on newer versions of Python and shouldn't need an upper bound

- https://pypi.org/project/django-guardian/#files

File: example_project/settings.py
Comment (ZbigniewRA) [reply example_project/settings.py]: Changed.

File: guardian/testapp/migrations/0006_auto_20230727_0658.py
Comment (ZbigniewRA) [reply guardian/testapp/migrations/0006_auto_20230727_0658.py:6]: I'm not sure what this is about, but I understand this is for later. Leaving it without change.

File: example_project/requirements.txt
Comment (ZbigniewRA) [reply example_project/requirements.txt:1]: Leaving it as is.

File: setup.py
Comment (ZbigniewRA) [reply setup.py:15]: Leaving it as is.

File: tox.ini
Comment (ZbigniewRA) [reply tox.ini:7]: Leaving it as is.

File: setup.py
Comment (ZbigniewRA) [reply setup.py]: Fixed.

File: tox.ini
Comment (ZbigniewRA) [reply tox.ini]: Removed.

File: setup.py
Comment (ZbigniewRA) [reply setup.py:31]: Leaving it as is.

**Review by thclark — APPROVED** [review]
Minor changes since my last review, happy

**Review by GabrielCastelo-31 — APPROVED** [review]

File: .github/workflows/tests.yml
Comment (johnthagen) [inline .github/workflows/tests.yml:13]:
```suggestion
    runs-on: ubuntu-24.04
```

So that it is pinned/reproducible?

File: .github/workflows/tests.yml
Comment (johnthagen) [inline .github/workflows/tests.yml:39]: I suggest we pin these to latest version instead of `latest` since then people can't go back in time and reproduce the same environment.

File: .github/workflows/tests.yml
Comment (johnthagen) [inline .github/workflows/tests.yml:45]: I suggest we pin these to latest version instead of `latest` since then people can't go back in time and reproduce the same environment.

File: .github/workflows/tests.yml
Comment (ZbigniewRA) [reply .github/workflows/tests.yml]: Pinned to `ubuntu-24.04`,

File: .github/workflows/tests.yml
Comment (ZbigniewRA) [reply .github/workflows/tests.yml]: Pinned to `mariadb:11.4.2`.

File: .github/workflows/tests.yml
Comment (ZbigniewRA) [reply .github/workflows/tests.yml]: Pinned to `postgres:16.3`.

**Comment by ZbigniewRA** [comment]
Tests I've run succeeded: [see here](https://github.com/RowAnalytics/django-guardian/actions/runs/10211907047)
But tests triggered by @thclark 15 hours ago failed ([see here](https://github.com/django-guardian/django-guardian/actions/runs/10159264626/job/28226071711)) with error: `operator does not exist: character varying = integer`
This is a mystery to me. Any ideas what happened?

**Comment by BonaFideIT** [comment]
Hi,

We would also like to express our interest in continuing the maintenance of django-guardian.

We also think that an upgrade to the current LTS version of django (4.2) makes a lot of sense and should be a top priority.

Since the drop of support for end of life versions has already been briefly discussed here, we would like to remind you that all versions below 4.2 are end of life, see https://www.djangoproject.com/download/#supported-versions.

So we would like to suggest that only versions >=4.2 should be supported and the others <4.2 should be dropped. The simplest thing would be to support only the supported LTS versions, which should result in even less maintenance effort.

Drop Support Suggestion:
- django 3.2
- django 4.1

Support Suggestion:
- django 4.2 (LTS) + python 3.8, 3.9, 3.10, 3.11, 3.12 (>=4.2.8)
- django 5.0 + python 3.10, 3.11, 3.12
- django 5.1 + python 3.10, 3.11, 3.12

https://docs.djangoproject.com/en/dev/faq/install/#what-python-version-can-i-use-with-django

**Comment by StephenChan** [comment]
I'm interested in seeing this PR merged as well.

@ZbigniewRA Pardon the mention, but I spent some time looking at those two CI runs you linked, and I noticed that the commit-trees used by each run (660125e failing, 55ca348 passing) have differences in shortcuts.py: https://github.com/django-guardian/django-guardian/compare/660125e..55ca348 

By using this PR's latest commit as a base and just applying those shortcuts.py differences, I replicated the failed run's errors: https://github.com/StephenChan/django-guardian/pull/2

The shortcuts.py differences don't seem to be related to this PR. Maybe they were from a different PR? In any case, it doesn't seem to be a concern for this PR's mergeability.

**Comment by BonaFideIT** [comment]
@StephenChan These changes to shortcuts.py are from https://github.com/django-guardian/django-guardian/pull/768.

**Comment by BonaFideIT** [comment]
You're right, I overlooked the changes from #812. These changes seem to be the problem why the tests are failing: https://github.com/django-guardian/django-guardian/actions/runs/9969776605/

**Comment by BonaFideIT** [comment]
We fixed the tests in https://github.com/django-guardian/django-guardian/pull/827.

**Review by BonaFideIT — CHANGES_REQUESTED** [review]
Looks good, but we would like to add changes from https://github.com/RowAnalytics/django-guardian/pull/3.

File: example_project/requirements.txt
Comment (michael-k) [reply example_project/requirements.txt:1]: `~=` already includes an upper bound.  Also: This an example project, not django-guardian's requirements.

File: .github/workflows/tests.yml
Comment (ZbigniewRA) [reply .github/workflows/tests.yml:18]: I don't know how it works. If you'll let me know how the `test.yml` file should look now I will fix it.

Comment (BonaFideIT) [reply .github/workflows/tests.yml:18]: We run the checks here: https://github.com/BonaFideIT/django-guardian/pull/5 and an it worked. @michael-k what exactly is the problem here?

**Review by BonaFideIT — APPROVED** [review]

File: .github/workflows/tests.yml
Comment (BonaFideIT) [reply .github/workflows/tests.yml:18]: I triggered the github checks and all checks have passed.

**Comment by BonaFideIT** [comment]
@brianmay and @thclark Can you please review this PR again?

File: .github/workflows/tests.yml
Comment (michael-k) [reply .github/workflows/tests.yml:18]: The tests won't fail, but they'll install the wrong version.  `Django~=4.1` means `Django>4.1,==4.*`, see https://peps.python.org/pep-0440/#compatible-release

Example: https://github.com/BonaFideIT/django-guardian/actions/runs/11381238190/job/31662202522?pr=5#step:6:54

Comment (michael-k) [reply .github/workflows/tests.yml:18]:
```suggestion
        django: ["3.2.25", "4.1.13", "4.2.16", "5.0.9", "main"]
```

File: .github/workflows/tests.yml
Comment (michael-k) [inline .github/workflows/tests.yml:21]: No need to exclude it.  Even if you put `4.2.0` in the matrix, with `~=` it won't install `4.2.[0-7]`.

```suggestion
```

File: .github/workflows/tests.yml
Comment (thclark) [reply .github/workflows/tests.yml:18]: I agree with this I think it's the last tweak before we merge. @ZbigniewRA if you could make this change please? Then we'll get this merged and make a release!!

Comment (BonaFideIT) [reply .github/workflows/tests.yml:18]: @michael-k Thanks for clarifying! Now we understand the problem and we completely agree with you.

File: .github/workflows/tests.yml
Comment (ZbigniewRA) [reply .github/workflows/tests.yml]: Removed as suggested.

File: .github/workflows/tests.yml
Comment (ZbigniewRA) [reply .github/workflows/tests.yml]: Changes to versions suggested by michael-k.

**Review by thclark — APPROVED** [review]

**Comment by thclark** [comment]
Given that this has been re-approved ad-infinitum and the only changes made since previous reviews have been to address their comments, I'm comfortable that there's enough consensus now to merge this. Thank you so much for your work everybody!

**Pull request merged**
